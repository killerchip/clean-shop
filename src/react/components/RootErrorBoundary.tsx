import { ErrorBoundaryProps } from "expo-router";
import { View } from "react-native";
import { Button, Text } from "react-native-paper";
import styled from "styled-components/native";
import * as Updates from "expo-updates";

type Props = Pick<ErrorBoundaryProps, "error"> & {
  testID?: string;
};

export function RootErrorBoundary({ error, testID }: Props) {
  const reload = () => Updates.reloadAsync();

  return (
    <Container testID={testID}>
      <Text variant="displayMedium">Oops!</Text>
      <Text variant="titleMedium">Something went wrong:</Text>
      <Text>{error.message}</Text>
      <ReloadButton mode="contained" onPress={reload}>
        Restart the App
      </ReloadButton>
    </Container>
  );
}

const Container = styled(View)`
  flex: 1;
  height: 200px;
  justify-content: center;
  align-items: center;
  padding: 20px;
`;

const ReloadButton = styled(Button)`
  margin-top: 20px;
`;
