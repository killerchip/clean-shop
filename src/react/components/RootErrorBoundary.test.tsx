import { render, fireEvent, screen } from "@testing-library/react-native";
import * as Updates from "expo-updates";
import { RootErrorBoundary } from "./RootErrorBoundary";

jest.mock("expo-updates");

describe("RootErrorBoundary Component", () => {
  const errorMessage = new Error("Test Error Message");

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("displays the error message", () => {
    render(<RootErrorBoundary error={errorMessage} />);
    expect(screen.getByText(errorMessage.message)).toBeTruthy();
  });

  it("renders with a testID when provided", () => {
    const testID = "error-boundary";
    render(<RootErrorBoundary testID={testID} error={errorMessage} />);
    expect(screen.getByTestId(testID)).toBeTruthy();
  });

  it("calls reloadAsync on Updates when Restart the App button is pressed", () => {
    render(<RootErrorBoundary error={errorMessage} />);
    fireEvent.press(screen.getByText("Restart the App"));
    expect(Updates.reloadAsync).toHaveBeenCalled();
  });

  it("matches snapshot", () => {
    render(<RootErrorBoundary error={errorMessage} />);
    expect(screen.toJSON).toMatchSnapshot();
  });
});
