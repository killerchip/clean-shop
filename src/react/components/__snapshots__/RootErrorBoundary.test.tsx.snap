// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RootErrorBoundary Component matches snapshot 1`] = `[Function]`;
